src/algorithms/convex.rs

```diff
@@ -0,0 +1,229 @@
+//! Convexity checking for portgraphs.
+//!
+//! This is based on a [`ConvexChecker`] object that is expensive to create
+//! (linear in the size of the graph), but can be reused to check multiple
+//! subgraphs for convexity quickly.
+
+use std::collections::{BTreeMap, BTreeSet};
+
+use crate::algorithms::toposort;
+use crate::{Direction, LinkView, NodeIndex, PortIndex};
+
+use super::TopoSort;
+
+#[derive(Default, Clone, Debug, PartialEq, Eq)]
+enum Causal {
+    #[default]
+    P, // in the past
+    F, // in the future
+}
+
+/// A pre-computed datastructure for fast convexity checking.
+///
+/// TODO: implement for graph traits?
+pub struct ConvexChecker<G> {
+    graph: G,
+    // The nodes in topological order
+    topsort_nodes: Vec<NodeIndex>,
+    // The index of a node in the topological order (the inverse of topsort_nodes)
+    topsort_ind: BTreeMap<NodeIndex, usize>,
+    // A temporary datastructure used during `is_convex`
+    causal: Vec<Causal>,
```

Comment:
Is this used to avoid allocations on each `is_node_convex` call? If not sure it's that necessary, specially since we'd only need to allocate `max_ind - min_ind` elems on each call (which hopefully is smaller than `n`).

Also, a `bitvec` would be better (or a `CausalFlags` wrapper to keep the nice `Causal` interface).

Reply:
I'd rather leave this option (because in my usecase I really expect a lot of calls to this), but I can offer a third option, which would be to pass this as an optional argument to `is_convex`. What do you think?

Reply:
Ahh but that would require to make `Causal` and `CausalVec` public, which I'm not too keen on.

Reply:
Makes sense. Just leave it and we can change it in the future.
(but check the `bitvec` option, `Vec<bool>` is quite inefficient).

Reply:
Now using `bitvec` 😊